packages/graphql-language-service-server/src/GraphQLCache.js

```diff
@@ -387,53 +372,40 @@ export class GraphQLCache {
     });
   }
 
-  getSchema = async (
-    configSchemaPath: ?Uri,
-    appName: ?string,
-  ): Promise<?GraphQLSchema> => {
-    if (!configSchemaPath) {
+  getSchema = async (appName: ?string): Promise<?GraphQLSchema> => {
+    const projectConfig = this._graphQLConfig.getProjectConfig(appName);
+
+    if (!projectConfig) {
       return null;
     }
-    const schemaPath = path.join(this._configDir, configSchemaPath);
-    if (this._schemaMap.has(schemaPath)) {
-      const schema = this._schemaMap.get(schemaPath);
-      return schema ? this._extendSchema(schema, schemaPath) : schema;
+    if (this._schemaMap.has(projectConfig.schemaPath)) {
+      const schema = this._schemaMap.get(projectConfig.schemaPath);
+      return schema
+        ? this._extendSchema(schema, projectConfig.schemaPath)
+        : schema;
     }
 
-    const schemaDSL = await new Promise(resolve =>
-      fs.readFile(schemaPath, 'utf8', (error, content) => {
-        if (error) {
-          throw new Error(error);
-        }
-        resolve(
-          content +
-            '\n' +
-            this._graphQLConfig.getCustomDirectives(appName).join('\n'),
-        );
-      }));
-
-    const schemaFileExt = path.extname(schemaPath);
-    let schema;
+    let schemaSDL;
```

Comment:
I'm worried of the performance of this approach to generate GraphQL schema. From what I understand, this approach will read the file, build `GraphQLSchema` object, print the schema in SDL format, and build it again. Last time I checked, `buildSchema()` takes around several seconds with 4~5k types available, and with `extendSchema()` being slow as well, this is definitely a hit with the performance.

I guess this is one of the ways to supprot `.json` format, which previously had bugs with my implementation to append the custom directives in SDL format to content string. One easy performance optimization we can make is to optionally perform custom directive extension when `customDirective` extension entry is available, with `extendSchema`, not `buildSchema`.